File: solutions/task_0438.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        if (int(s.size()) < int(p.size())) {
            return {};
        }
        vector<int> cur(30);
        vector<int> shouldBe(30);
        for (int i = 0; i < int(p.size()); i++) {
            cur[s[i]-'a'] += 1;
            shouldBe[p[i]-'a'] += 1;
        }
        vector<int> ans;
        if (cur == shouldBe) {
            ans.push_back(0);
        }
        int lnp = int(p.size());
        for (int i = lnp; i < int(s.size()); i++) {
            cur[s[i-lnp]-'a'] -= 1;
            cur[s[i]-'a'] += 1;
            if (cur == shouldBe) {
                ans.push_back(i-lnp+1);
            }
        }
        return ans;
    }
};

File: solutions/task_0438.py
from collections import Counter

class Solution:
    def findAnagrams(self, s: str, p: str):
        if len(s) < len(p):
            return []
        shouldBe = 26 * [0]
        state = 26 * [0]
        for i in range(len(p)):
            shouldBe[ord(p[i])-ord('a')] += 1
            state[ord(s[i])-ord('a')] += 1
        ans = []
        if state == shouldBe:
            ans.append(0)
        for i in range(len(p), len(s)):
            state[ord(s[i-len(p)])-ord('a')] -= 1
            state[ord(s[i]) - ord('a')] += 1
            if state == shouldBe:
                ans.append(i-len(p)+1)
        return ans
